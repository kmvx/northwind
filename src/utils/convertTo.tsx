export type DataType = Record<string, any>[] | undefined;

const getHeaders = (data: Record<string, any>[]): string[] => {
  //return Object.keys(data[0]);
  return Array.from(new Set(data.flatMap(Object.keys)));
};

const isDateString = (value: string): boolean => {
  return (
    /^\d{4}-\d{2}-\d{2}(T\d{2}:\d{2}:\d{2}.*)?$/.test(value) &&
    !isNaN(Date.parse(value))
  );
};

const formatDate = (date: Date): string => {
  if (date.getHours() || date.getMinutes() || date.getSeconds()) {
    return date.toLocaleString();
  } else {
    return date.toLocaleDateString();
  }
};

const formatValue = (value: any): string => {
  if (value === null || value === undefined) {
    return '';
  }

  if (value instanceof Date) {
    return formatDate(value);
  }

  if (typeof value === 'string' && isDateString(value)) {
    return formatDate(new Date(value));
  }

  return String(value);
};

export const convertToCSV = (data: DataType): string => {
  if (!data?.length) return '';

  const headers = getHeaders(data);

  const escapeCSV = (val: string): string => {
    return /[,"\n\r]/.test(val) ? `"${val.replace(/"/g, '""')}"` : val;
  };

  const rows = data.map((row) => {
    return headers
      .map((header) => {
        const val = row[header];

        if (val === null || val === undefined) {
          return '';
        }

        if (typeof val === 'number' || typeof val === 'boolean') {
          return String(val);
        }

        if (val instanceof Date) {
          return `"${val.toISOString()}"`;
        }

        return escapeCSV(String(val));
      })
      .join(',');
  });

  return [headers.join(','), ...rows].join('\n');
};

export const convertToMarkdown = (data: DataType, title: string): string => {
  if (!data?.length) return '';

  const headers = getHeaders(data);

  const escapeMarkdown = (val: string): string => {
    return val.replace(/\|/g, '\\|');
  };

  const formatted = (value: any): string => {
    return escapeMarkdown(formatValue(value));
  };

  const columnWidths = headers.map((header) => {
    const cellWidths = data.map((row) => formatted(row[header]).length);
    return Math.max(header.length, ...cellWidths);
  });

  const pad = (str: string, len: number): string => {
    return str + ' '.repeat(len - str.length);
  };

  const formatRow = (cells: string[]): string => {
    return `| ${cells.map((cell, i) => pad(cell, columnWidths[i])).join(' | ')} |`;
  };

  const headerRow = formatRow(headers);
  const separatorRow = `| ${columnWidths.map((w) => '-'.repeat(w)).join(' | ')} |`;
  const dataRows = data.map((row) => {
    return formatRow(headers.map((header) => formatted(row[header])));
  });

  return [`# ${title}`, '', headerRow, separatorRow, ...dataRows].join('\n');
};

export const convertToJSON = (data: DataType): string => {
  return JSON.stringify(data, null, 4);
};
