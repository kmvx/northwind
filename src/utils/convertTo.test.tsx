import { expect, it } from 'vitest';
import { convertToCSV, convertToJSON, convertToMarkdown } from './convertTo';

it('convertTo', () => {
  const data = [
    {
      number: 1.2,
      dateObject: new Date(Date.UTC(2025)),
      dateTimeObject: new Date(Date.UTC(2025, 2, 3, 10, 5, 6, 999)),
      stringDate: '1992-08-14T00:00:00',
      invalidDate: new Date(NaN),
      time: '2025-06-14T18:54:09',
    },
  ];
  expect(convertToCSV(data)).toMatchSnapshot();
  expect(convertToMarkdown(data, 'Datas')).toMatchSnapshot();
  expect(convertToJSON(data)).toMatchSnapshot();
});
