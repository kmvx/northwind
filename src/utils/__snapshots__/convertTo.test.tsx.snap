// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convertTo 1`] = `
"number,dateObject,dateTimeObject,stringDate,invalidDate,time
1.2,2025-01-01T00:00:00.000Z,2025-03-03T10:05:06.999Z,1992-08-14T00:00:00,,2025-06-14T18:54:09"
`;

exports[`convertTo 2`] = `
"# Datas

| number | dateObject           | dateTimeObject        | stringDate | invalidDate  | time                  |
| ------ | -------------------- | --------------------- | ---------- | ------------ | --------------------- |
| 1.2    | 1/1/2025, 2:00:00 AM | 3/3/2025, 12:05:06 PM | 8/14/1992  | Invalid Date | 6/14/2025, 6:54:09 PM |"
`;

exports[`convertTo 3`] = `
"[
    {
        "number": 1.2,
        "dateObject": "2025-01-01T00:00:00.000Z",
        "dateTimeObject": "2025-03-03T10:05:06.999Z",
        "stringDate": "1992-08-14T00:00:00",
        "invalidDate": null,
        "time": "2025-06-14T18:54:09"
    }
]"
`;
